### DISCUSSION TITLE
How can I implement COM interface imported with cswin32 without Marshaling for AOT?

### DISCUSSION THREAD

**Discussion by ghord** [microsoft/CsWin32#1236] [Q&A]
I've been trying to implement COM interface (specifically [ITextHost](https://learn.microsoft.com/en-us/windows/win32/api/textserv/nl-textserv-itexthost)) from my C# code in a way that is AOT friendly.
I was able to manually declare this interface with all methods and implement it using COM Wrappers code generator approach, but I was wondering whether this is possible with cswin32 currently, so that I do not have to code method signatures by hand?

When I import `ITextHost` interfaces without Marshaling (required for AOT), is there some simple way I can implement it on my class that will be passed to code expecting `ITextHost`? Currently code generated by cswin32 is a `struct`, I don't see any mentions how would I implement this interface. I've read mentions that cswin32 does use COM Wrapper compatible code when Marshaling is disabled, but generated code is really different than what is mentioned in [COM Wrappers source generator article](https://learn.microsoft.com/en-us/dotnet/standard/native-interop/comwrappers-source-generation).

Right now I have to use two different approaches for COM interop - cswin32 for importing interfaces and COM Wrappers source generator for exposing interfaces, which I would like to unify.

I've noticed that generated code includes `ITextHost.Interface` interface, but it is not partial so I cannot annotate it with `[GeneratedComInterface]` attribute nor can I change calling conventions of methods to ThisCall (ITextHost has weird calling convention). I'm not sure how I would do `QueryInterface` on an object implementing `ITextHost.Interface` (this is needed for CreateTextServices API call).

I've tried to copy generated interface and simply annotate it with attributes for COM Wrapper generator, but compilation errors out with marshalling failures such as:
`The type 'Windows.Win32.Foundation.HRESULT' is not supported by source-generated COM. The generated source will not handle marshalling of the return value of method 'TxGetBackStyle'. `
`For types that are not supported by source-generated COM, the resulting function pointer will rely on the underlying runtime to marshal the specified type.`

I've also noticed that COM Wrappers api has strategy based methods for managing lifetime of instances of COM objects, how does this relate to cswin32?

**Comment by AArnott** [comment]
Great questions. We hope to get some good guidance and probably some cswin32 changes made in the next month or so to improve the story here. 

In the meantime, @JeremyKuhne may have some ideas to help as he already does a lot of trim-friendly COM stuff with cswin32 in WinForms.
